eng/targets/ResolveReferences.targets

```diff
@@ -148,6 +148,60 @@
            Text="Could not resolve this reference. Could not locate the package or project for &quot;%(Reference.Identity)&quot;" />
   </Target>
 
+  <!--
+    Remove @(ResolvedCompileFileDefinitions) and @(RuntimeCopyLocalItems) items for patched assemblies between item
+    generation and use. Instead, ensure @(ProjectReference) items include the patched projects. This covers indirect
+    references and improves test coverage; @(ProjectReference) items were set correctly for direct references. Changes
+    are necessary only in servicing builds, in the release/2.1 branch, and when building non-implementation projects.
+
+    This is imperfect because it covers only indirect references to project reference providers (most C#
+    implementation projects) and does not work for legacy samples or other non-.NET SDK projects.
+
+    Note: Project reference providers have matching assembly and project names. %(Filename) matches work :)
+  -->
+  <Target Name="_TestWithPatchedAssemblies"
+      Condition=" $(UseProjectReferences) AND '$(PackagesInPatch)' != '' AND
+          (@(ResolvedCompileFileDefinitions->Count()) != 0 OR @(RuntimeCopyLocalItems->Count()) != 0) "
+      AfterTargets="ResolvePackageAssets"
+      BeforeTargets="AssignProjectConfiguration;
+          GenerateBuildDependencyFile;
+          GeneratePublishDependencyFile;
+          ILLink;
+          ResolveLockFileCopyLocalFiles;
+          ResolveLockFileReferences"
+      DependsOnTargets="ResolvePackageAssets">
+    <!-- Find projects patched in this build that need to be added as references. -->
+    <ItemGroup>
+      <_ProjectsInPatch Remove="@(_ProjectsInPatch)" />
+      <_ProjectsInPatch Include="@(ProjectReferenceProvider->'%(ProjectPath)')"
+          Condition=" $(PackagesInPatch.Contains(' %(ProjectReferenceProvider.Identity);')) "
+          KeepMetadata="None" />
+      <!-- dotnet-razorpagegenerator project is unique in overriding its package ID. -->
+      <_ProjectsInPatch Include="@(ProjectReferenceProvider->WithMetadataValue('Identity', 'dotnet-razorpagegenerator')->'%(ProjectPath)')"
+          Condition=" $(PackagesInPatch.Contains(' RazorPageGenerator;')) "
+          KeepMetadata="None" />
+
+      <!-- Ignore unreferenced assemblies and directly-referenced projects. -->
+      <_ProjectsToAdd Remove="@(_ProjectsToAdd)" />
+      <!-- Condition uses same Filename metadata when evaluating the other two item groups. -->
+      <_ProjectsToAdd Include="@(_ProjectsInPatch)"
+          Condition=" '%(Filename)' != '' AND
```

Comment:
What is this condition doing exactly? Add a direct reference to the projects being patched, unless...

Reply:
My mental model may be wrong (@Forgind and @yuehuang010, please chime in) but https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-batching?view=vs-2019#item-and-property-mutations contains _some_ of the information
1. The `'%(Filename)` item metadata causes the implicit `CreateItem` task to be batched
2. Use of that metadata also limits the `@(ResolvedCompileFileDefinitions)` and `@(RuntimeCopyLocalItems)` item groups to the item(s) with the same `%(Filename)`
    - This part doesn't seem to be explicit in the documentation
3. The `OR` clause confirms at least one such item exists

At a higher level, this creates a `@(_ProjectsToAdd)` item group containing those `@(_ProjectsInPatch)` items that match on filename with  one or more `@(ResolvedCompileFileDefinitions)` items `@(RuntimeCopyLocalItems)`, excluding existing `@(ProjectReference)` items. Another way of looking at it is the created item group contains the projects in the current patch that (a) aren't referenced directly but (b) match assemblies in the current build.

Reply:
I see it used in this exampled (https://docs.microsoft.com/en-us/visualstudio/msbuild/item-metadata-in-task-batching?view=vs-2019#divide-several-item-lists-into-batches).  So, it is feature name is "Divide several item lists into batches":
``````
    <Target Name="ShowMessage">
        <Message
            Text = "Number: %(Number) -- Items in ExampColl: @(ExampColl) ExampColl2: @(ExampColl2)"/>
    </Target>
``````

Reply:
Good point @yuehuang010. I'm not sure how many times I read that example without thinking "Oh, this makes intersection / subsetting dead easy"😃